src/Components/Aspire.Keycloak/AspireKeycloakExtensions.cs

```diff
@@ -0,0 +1,117 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Keycloak;
+using Microsoft.AspNetCore.Authentication.Cookies;
+using Microsoft.AspNetCore.Authentication.JwtBearer;
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Extensions.Hosting;
+
+/// <summary>
+/// Provides extension methods for registering Keycloak-related services in an <see cref="IHostApplicationBuilder"/>.
+/// </summary>
+public static class AspireKeycloakExtensions
+{
+    private const string DefaultConfigSectionName = "Aspire:Keycloak";
+    private const string KeycloakBackchannel = nameof(KeycloakBackchannel);
+
+    /// <summary>
+    /// Adds Keycloak JWT Bearer authentication to the application.
+    /// </summary>
+    /// <param name="builder">The <see cref="IHostApplicationBuilder" /> to read config from and add services to.</param>
+    /// <param name="connectionName">The connection name to use to find a connection string.</param>
+    /// <param name="authenticationScheme">The authentication scheme name. Default is "Bearer".</param>
+    /// <param name="configureJwtBearerOptions">An optional action to configure the <see cref="JwtBearerOptions"/>.</param>
+    /// <param name="configureSettings">An optional action to configure the <see cref="KeycloakSettings"/>.</param>
+    public static void AddKeycloakJwtBearer(
+        this IHostApplicationBuilder builder,
+        string connectionName,
+        string authenticationScheme = JwtBearerDefaults.AuthenticationScheme,
+        Action<JwtBearerOptions>? configureJwtBearerOptions = null,
+        Action<KeycloakSettings>? configureSettings = null)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        var settings = new KeycloakSettings();
+        builder.Configuration.GetSection(DefaultConfigSectionName).Bind(settings);
+
+        if (builder.Configuration.GetConnectionString(connectionName) is string connectionString)
+        {
+            settings.ParseConnectionString(connectionString);
+        }
+
+        configureSettings?.Invoke(settings);
+
+        builder.Services.AddHttpClient(KeycloakBackchannel, o => o.BaseAddress = settings.Endpoint);
+
+        builder.Services.AddAuthentication(authenticationScheme)
+                .AddJwtBearer(authenticationScheme);
+
+        builder.Services
+               .AddOptions<JwtBearerOptions>(authenticationScheme)
+               .Configure<IConfiguration, IHttpClientFactory, IHostEnvironment>((options, configuration, httpClientFactory, hostEnvironment) =>
+               {
+                   var backchannelHttpClient = httpClientFactory.CreateClient(KeycloakBackchannel);
+
+                   options.Backchannel = backchannelHttpClient;
+                   options.Authority = $"{backchannelHttpClient.BaseAddress}/realms/{settings.Realm}";
```

Comment:
Damian, would be great to have the realm configured in the AppHost. But what's the exact expectation here?

Like this?

```csharp
var keycloak = builder.AddKeycloak("keycloak", realm: "myrealm" );
```

Or like this?

```csharp
var keycloak = builder.AddKeycloak("keycloak")
                      .WithRealm("myrealm");
```

Or like this?

```csharp
var apiService = builder.AddProject<Projects.Keycloak_ApiService>("apiservice")
                        .WithReference(keycloak, "myrealm");
```

Happy to implement it either way, but could you please point me to an example I can follow of the right way to do it?

Reply:
Is Realm required (meaning is there a default if none provided)? If not, then other code building up the URI is going to eventually fail as you set the Authority, correct?

Realm to me kinda feels like a database pattern to me like you add the resource, then the realm (AddPostgres().AddDatabase)) - I kinda think this would be `AddKeyCloak("keycloak").AddRealm("myrealm")`

Reply:
I was thinking the last example `.WithReference(keycloak, "myrealm")`, rather than promote realms to child resources of their own, although I see the similarity to databases vs. database servers. Interested in others' thoughts @davidfowl @mitchdenny @eerhardt

Reply:
@DamianEdwards @timheuer Two comments on this one:

1. Is adding the realm on the host what we really want? The realm is like an AAD tenant and that is something that you usually configure on your consuming projects. Feels a bit too magical for client projects to have no idea what realm they are working with.

2. Assuming we set the realm on the host, what's the expectation of `WithReference(keycloak, "myrealm")`? I don't know of a way to construct a discoverable endpoint like http://localhost:1234/realms/WeatherShop directly in the host side. But I could set an environment variable that the client component will know about and read to build the endpoint. Would that work?

Reply:
> Feels a bit too magical for client projects to have no idea what realm they are working with.

The realm has to exist on the Keycloak server though. Client projects get the database details they're connecting to as well so not sure this is much different.

> But I could set an environment variable that the client component will know about and read to build the endpoint. Would that work?

Yep that's what I was thinking. This again is very similar to what happens when you call `AddDatabase("somename")`. The database isn't automatically created, but it gets added to the connection information for any client project calling `WithReference` on it.

Reply:
Added the WithReference overload. In the host you can do this now:

```csharp
builder.AddProject<Projects.Keycloak_ApiService>("apiservice")
       .WithReference(keycloak, realm: "WeatherShop");
```

And client projects can do just this, since host provides everything:

```csharp
builder.AddKeycloakJwtBearer();
```

Reply:
Now client projects need to pass in the connection name like this:

```csharp
builder.AddKeycloakJwtBearer("keycloak");
```

Or this:

```csharp
builder.AddKeycloakOpenIdConnect("keycloak");
```